File: examples/python/src/money.py
"""Whole-cent money helpers, the Python twin of the JavaScript example.

Amounts are integers throughout; a float never touches a price.
"""

CENTS_PER_UNIT = 100
MAX_PARTS = 12


def from_units(units: int) -> int:
    if not isinstance(units, int):
        raise TypeError("units must be an integer count of cents")
    return units


def units_to_text(units: int) -> str:
    sign = "-" if units < 0 else ""
    magnitude = abs(units)
    whole = magnitude // CENTS_PER_UNIT
    part = magnitude % CENTS_PER_UNIT
    return f"{sign}{whole}.{part:02d}"


def add_all(amounts):
    total = 0
    for amount in amounts:
        total += from_units(amount)
    return total


def split_evenly(units: int, parts: int):
    if parts < 1 or parts > MAX_PARTS:
        raise ValueError("parts must be between 1 and 12")
    each = units // parts
    rest = units - each * parts
    out = []
    for index in range(parts):
        extra = 1 if index < rest else 0
        out.append(each + extra)
    return out


def percent_of(units: int, percent: float) -> int:
    return int(round(units * percent / 100))

File: examples/javascript/src/money.js
"use strict";

// Whole-unit money arithmetic. Every amount in this project is an integer count
// of the smallest unit (cents, fen, sen), so no rounding ever happens here.

const CENTS_PER_UNIT = 100;
const MAX_PARTS = 12;

function fromUnits(units) {
  if (!Number.isInteger(units)) {
    throw new Error("units must be an integer count of cents");
  }
  return units;
}

function unitsToText(units) {
  const sign = units < 0 ? "-" : "";
  const magnitude = Math.abs(units);
  const whole = Math.floor(magnitude / CENTS_PER_UNIT);
  const part = magnitude % CENTS_PER_UNIT;
  if (part < 10) {
    return sign + whole + ".0" + part;
  }
  return sign + whole + "." + part;
}

function addAll(amounts) {
  let total = 0;
  for (let i = 0; i < amounts.length; i += 1) {
    total += fromUnits(amounts[i]);
  }
  return total;
}

/// Split `units` into `parts` whole units, giving the remainder to the first
/// lines so the split always sums back to the original amount.
function splitEvenly(units, parts) {
  if (parts < 1 || parts > MAX_PARTS) {
    throw new Error("parts must be between 1 and 12");
  }
  const each = Math.floor(units / parts);
  const rest = units - each * parts;
  const out = [];
  for (let i = 0; i < parts; i += 1) {
    out.push(each + (i < rest ? 1 : 0));
  }
  return out;
}

function percentOf(units, percent) {
  return Math.round((units * percent) / 100);
}

module.exports = {
  CENTS_PER_UNIT,
  fromUnits,
  unitsToText,
  addAll,
  splitEvenly,
  percentOf,
};
